libaugrim/src/algorithm/two_phase_commit/participant_context.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021-2022 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::process::Process;
+use crate::time::Time;
+
+use super::Epoch;
+
+#[derive(Clone)]
+pub enum ParticipantState<T>
+where
+    T: Time,
+{
+    Abort,
+    Commit,
+    Voted {
+        vote: bool,
+        decision_timeout_start: T,
+    },
+    WaitingForVoteRequest,
+    WaitingForVote,
+}
+
+#[derive(Clone)]
+pub struct ParticipantContext<P, T>
+where
+    P: Process,
+    T: Time,
+{
+    alarm: Option<T>,
+    epoch: Epoch,
+    coordinator: P,
+    last_commit_epoch: Option<Epoch>,
+    participant_processes: Vec<P>,
+    state: ParticipantState<T>,
+    this_process: P,
+}
+
+impl<P, T> ParticipantContext<P, T>
+where
+    P: Process,
+    T: Time,
+{
+    pub fn new(this_process: P, coordinator: P, participant_processes: Vec<P>) -> Self {
```

Comment:
Based on the algorithm being followed, the Participant doesn't necessarily know the other participants in advance.

Reply:
This is an intentional deviation. CCRDS suggests a possible different approach (passing participants during VoteRequest), and if we switch to that approach we will want to update the coordinator algorithm/messages at the same time.